datafusion/common/src/types/native.rs

```diff
@@ -253,6 +251,8 @@ impl LogicalType for NativeType {
             (Self::Interval(iu), _) => Interval(*iu),
             (Self::Binary, LargeUtf8) => LargeBinary,
             (Self::Binary, Utf8View) => BinaryView,
+            // We don't cast to another kind of binary type if the origin one is already a binary type
+            (Self::Binary, Binary | LargeBinary | BinaryView) => origin.to_owned(),
```

Comment:
Same thing as we do for strings:

https://github.com/apache/datafusion/blob/fad5f8a628acc7db29ffed2c3faf7f1f35da5837/datafusion/common/src/types/native.rs#L266-L267

Looks like we just missed it for binary types.